core/src/main/java/org/apache/iceberg/CachingCatalog.java

```diff
@@ -30,23 +35,98 @@
+
+  private static final Logger LOG = LoggerFactory.getLogger(CachingCatalog.class);
+  private static final RemovalListener<TableIdentifier, Table> identLoggingRemovalListener =
+      (key, value, cause) -> LOG.debug("{} was evicted from the TableCache with {} cause", key, cause);
+
+  // TODO - Update this to use the default expiration interval millis once things are more complete.
+  //        Don't want to break current tests.
   public static Catalog wrap(Catalog catalog) {
-    return wrap(catalog, true);
+    return wrap(catalog, 0);
+  }
+
+  public static Catalog wrap(Catalog catalog, long expirationIntervalMillis) {
+    return wrap(catalog, true, expirationIntervalMillis);
+  }
+
+  public static Catalog wrap(Catalog catalog, Duration expirationInterval) {
+    return wrap(catalog, true, expirationInterval.toMillis());
   }
 
-  public static Catalog wrap(Catalog catalog, boolean caseSensitive) {
-    return new CachingCatalog(catalog, caseSensitive);
+  public static Catalog wrap(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis) {
+    return new CachingCatalog(catalog, caseSensitive, expirationIntervalMillis);
   }
 
-  private final Cache<TableIdentifier, Table> tableCache = Caffeine.newBuilder().softValues().build();
   private final Catalog catalog;
   private final boolean caseSensitive;
+  private final boolean expirationEnabled;
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final long expirationIntervalMillis;
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected final Cache<TableIdentifier, Table> tableCache;
+
+  private CachingCatalog(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis) {
+    this(catalog, caseSensitive, expirationIntervalMillis, Ticker.systemTicker());
+  }
 
-  private CachingCatalog(Catalog catalog, boolean caseSensitive) {
+  @SuppressWarnings("checkstyle:VisibilityModifier")
+  protected CachingCatalog(Catalog catalog, boolean caseSensitive, long expirationIntervalMillis, Ticker ticker) {
     this.catalog = catalog;
     this.caseSensitive = caseSensitive;
+    // Set negative values to zero to avoid creating a Duration with a negative value
+    this.expirationIntervalMillis = expirationIntervalMillis <= 0 ? 0 : expirationIntervalMillis;
+    this.expirationEnabled = expirationIntervalMillis > 0;
+    this.tableCache = createTableCache(ticker);
+  }
+
+  private Cache<TableIdentifier, Table> createTableCache(Ticker ticker) {
+    Caffeine<TableIdentifier, Table> cacheBuilder = Caffeine
+        .newBuilder()
+        .softValues()
+        .removalListener(identLoggingRemovalListener);
+
+    // Only setup ticker and background cleanup tasks from expiration if table expiration is enabled.
+    if (expirationEnabled) {
+      return cacheBuilder
+          .writer(new CacheWriter<TableIdentifier, Table>() {
+            @Override
+            public void write(TableIdentifier tableIdentifier, Table table) {
+            }
```

Comment:
It may be cleaner for this to be an inner class (not static).

Reply:
The preference is really to inline them. The compiler has trouble figuring out the types otherwise.

Almost all examples I've seen have inlined like this. But I'll see if I can get it work. Right now it's complaining because the `Caffeine` object generated is really `Caffeine<Object, Object>` for a long time and they have some weird tricks that do the type inferring after a bit. Hence the preference towards in-lining and a fluent builder where posible.

Reply:
Either way.

Reply:
I managed to get it. Upgrading my IntelliJ caused a lot of headaches. I downgraded and it seems fine now.